### PR TITLE
Add TreasuryV3

### PR DESCRIPTION
**Pull request by professor-boost**

### CODE REVIEW

File: contracts/TreasuryV3.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    function setSwapRouter(SwapRouter _swapRouter) external onlyGov {
+        swapRouter = _swapRouter;
+    }
+
+    function setOrbitStation(address _orbitStation) external onlyGov {
+        orbitStation = _orbitStation;
+    }
+
+    function balanceOf(IERC20 token) public view returns (uint256) {
+        return token.balanceOf(address(this)).sub(ecoFundAmts[address(token)]);
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:72]: remove ```.sub(ecoFundAmts[address(token)])```

File: contracts/TreasuryV3.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function deposit(IERC20 token, uint256 amount) external {
+        token.safeTransferFrom(msg.sender, address(this), amount);
+    }
+
+    function withdrawERC20(uint256 amount, address withdrawAddress, address token) external onlyGov {
+        IERC20 coin = IERC20(token);
+        require(balanceOf(coin) >= amount, "insufficient funds");
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:81]: can remove this require statement, safeTransfer should handle it

File: contracts/TreasuryV3.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        coin.safeTransfer(withdrawAddress, amount);
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:82]: refactor to ```IERC20(token).safeTransfer(withdrawAddress, amount)```

File: contracts/TreasuryV3.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    function convertToOrbit(address[] calldata routeDetails, uint256 amount) external {
+        require(routeDetails[0] != address(nativeToken), "src can't be boost");
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:86]: remove, should allow for the src to be the ```BoostToken```

File: contracts/TreasuryV3.sol [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        require(routeDetails[routeDetails.length - 1] == address(nativeToken), "dest not defaultToken");
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:87]: “dest cannot be nativeToken”

File: contracts/TreasuryV3.sol [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        IERC20 srcToken = IERC20(routeDetails[0]);
+        require(balanceOf(srcToken) >= amount, "insufficient funds");
+        if (srcToken.allowance(address(this), address(swapRouter)) <= amount) {
+            srcToken.safeApprove(address(swapRouter), 0);
```
Comment (JoaquinLiHua) [inline contracts/TreasuryV3.sol:91]: should not need to safeApprove twice

**Review by JoaquinLiHua — CHANGES_REQUESTED** [review]
Left some comments

File: contracts/TreasuryV3.sol [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            amount,
+            0,
+            routeDetails,
+            address(this),
+            block.timestamp + 100
+        );
+    }
+
+
+    function fundOrbitStation(uint256 amount) external onlyGov {
```
Comment (professor-boost) [inline contracts/TreasuryV3.sol:100]: Potentially needs to call `notifyRewards` or update rewards on OrbitStation when Orbit proceeds sent to staking contract

**Pull request merged**